File: src/main/scala/sync.scala
import de.vandermeer.asciitable.AsciiTable
import de.vandermeer.skb.interfaces.transformers.textformat.TextAlignment

import java.sql.{Connection, ResultSet}

def getAllSync(conn: Connection): Unit = {
  val sql = "SELECT * FROM sync"

  val stmt = conn.prepareStatement(sql)
  val res: ResultSet = stmt.executeQuery()

  val at = new AsciiTable()
  at.addRule()
  val row = at.addRow("NOME", "DIRECTORY", "SERVER", "PATH", "TYPE")
  row.setTextAlignment(TextAlignment.CENTER)
  at.addRule()

  while ( {
    res.next
  }) {
    val name = res.getString("name")
    val directory = res.getString("directory")
    val server = res.getString("server")
    val path = res.getString("path")
    val typeSync = res.getString("type")

    at.addRow(name, directory, server, path, typeSync)
    at.addRule()
  }

  val table = at.render()
  println(table)

  stmt.close()
}

def getSyncByName(conn: Connection, name: String): (Int, String) = {
  var strName = "";
  var count = 0;
  val sql = "SELECT * FROM sync WHERE name = ?"

  val stmt = conn.prepareStatement(sql)
  stmt.setString(1, name)

  val res: ResultSet = stmt.executeQuery()

  if (res.next()) {
    strName = res.getString("name")
    count += count + 1
  }

  stmt.close()

  (count, strName)
}

def addSync(conn: Connection, name: String, dir: String, server: String, path: String, typeSync: String): Unit = {
  val sql =
    """
      INSERT INTO sync (name, directory, server, path, type)
      VALUES (?, ?, ?, ?, ?)
      ON CONFLICT (name) DO UPDATE SET directory = ?, server = ?, path = ?, type = ?
    """

  val stmt = conn.prepareStatement(sql)
  stmt.setString(1, name)
  stmt.setString(2, dir)
  stmt.setString(3, server)
  stmt.setString(4, path)
  stmt.setString(5, typeSync)
  // UPDATE
  stmt.setString(6, dir)
  stmt.setString(7, server)
  stmt.setString(8, path)
  stmt.setString(9, typeSync)

  stmt.executeUpdate()
  stmt.close()
}

def delSync(conn: Connection, name: String): Unit = {
  val sql =
    """
      DELETE FROM sync WHERE name = ?
    """

  val stmt = conn.prepareStatement(sql)
  stmt.setString(1, name)

  stmt.executeUpdate()
  stmt.close()
}

File: rust_ftp_bck/src/sync.rs
use duckdb::Connection;
use prettytable::{color, Attr, Cell, Row, Table};
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
pub(crate) struct MySync {
  pub(crate) key: String,
  pub(crate) value: String,
  pub(crate) server: String,
  pub(crate) path: String,
}

static DB_DIR: &'static str = "DATABASE";
static DB_FILE: &'static str = "db.duckdb";

pub fn get_all_sync() {
  let db_path: PathBuf = Path::new(DB_DIR).join(DB_FILE);
  let conn = Connection::open(db_path).unwrap();

  let mut stmt = conn
      .prepare("SELECT key, value, server, path FROM sync ORDER BY key ASC")
      .unwrap();

  let rows = stmt
      .query_map([], |row| {
        let key: String = row.get(0)?;
        let value: String = row.get(1)?;
        let server: String = row.get(2)?;
        let path = row.get(3)?;

        Ok(MySync { key, value, server, path })
      })
      .unwrap();

  let mut table = Table::new();
  table.add_row(Row::new(vec![
    Cell::new("NOME")
        .with_style(Attr::Bold)
        .with_style(Attr::ForegroundColor(color::BLUE)),
    Cell::new("CARTELLA")
        .with_style(Attr::Bold)
        .with_style(Attr::ForegroundColor(color::CYAN)),
    Cell::new("SERVER")
        .with_style(Attr::Bold)
        .with_style(Attr::ForegroundColor(color::MAGENTA)),
    Cell::new("PATH")
        .with_style(Attr::Bold)
        .with_style(Attr::ForegroundColor(color::GREEN)),
  ]));

  for row in rows {
    let sync = row.unwrap();

    table.add_row(Row::new(vec![
      Cell::new(&sync.key),
      Cell::new(&sync.value),
      Cell::new(&sync.server),
      Cell::new(&sync.path),
    ]));
  }

  table.printstd();
  println!("=====");
}

pub fn get_sync_by_key(k: &str) -> Result<MySync, String> {
  let db_path: PathBuf = Path::new(DB_DIR).join(DB_FILE);
  let conn = Connection::open(db_path)
      .map_err(|e| e.to_string())?;

  let mut stmt = conn
      .prepare("SELECT key, value, server, path FROM sync WHERE key = ?")
      .map_err(|e| e.to_string())?;

  let mut rows = stmt.query([k])
      .map_err(|e| e.to_string())?;

  if let Some(row) = rows.next().map_err(|e| e.to_string())? {
    Ok(MySync {
      key: row.get(0).map_err(|e| e.to_string())?,
      value: row.get(1).map_err(|e| e.to_string())?,
      server: row.get(2).map_err(|e| e.to_string())?,
      path: row.get(3).map_err(|e| e.to_string())?,
    })
  } else {
    Err("Non è stato trovato alcun record".to_string())
  }
}

pub fn add_sync(key: &str, value: &str, server: &str, path: &str) {
  let db_path: PathBuf = Path::new(DB_DIR).join(DB_FILE);
  let conn = Connection::open(db_path).unwrap();

  conn.execute(
    "INSERT INTO sync (key, value, server, path) VALUES (?, ?, ?, ?) ON CONFLICT DO UPDATE SET value = ?, server = ?, path = ?",
    [
      // INSERT
      key,
      value,
      server,
      path,
      // UPDATE
      value,
      server,
      path,
    ],
  )
      .expect("ERRORE DI INSERIMENTO NELLA TABELLA sync");

  println!("OPERAZIONE AVVENUTA CON SUCCESSO!");
  println!("=====");
}

pub fn del_sync(key: &str) {
  let db_path: PathBuf = Path::new(DB_DIR).join(DB_FILE);
  let conn = Connection::open(db_path).unwrap();

  conn.execute(
    "DELETE FROM sync WHERE key = ?",
    [key],
  )
      .expect("ERRORE DI CANCELLAZIONE NELLA TABELLA ftp");

  println!("OPERAZIONE AVVENUTA CON SUCCESSO!");
  println!("=====");
}